### PR TITLE
mockkubeapiserver: Support stringData when creating a secret

### PR DESCRIPTION
**Pull request by justinsb**
This is an edge case in the kube apiserver, but there is special
handling for the stringData field of a secret, that is mapped to
base64 data.

### CODE REVIEW

**Comment by justinsb** [comment]
Hmmm .... this no longer needs a rebase (at least it shouldn't)...  I'll try a retest

/retest

File: mockkubeapiserver/postresource.go [hunk 1]
```diff
@@ -73,8 +74,42 @@ func (req *postResource) Run(ctx context.Context, s *MockKubeAPIServer) error {
...
 }
+
+func preCreateHook(ctx context.Context, obj *unstructured.Unstructured) error {
+	gvk := obj.GroupVersionKind()
+	if gvk.Group == "" && gvk.Version == "v1" && gvk.Kind == "Secret" {
+		stringDataAny, found := obj.Object["stringData"]
+		if found {
+			stringDataMap, ok := stringDataAny.(map[string]any)
+			if !ok {
+				return fmt.Errorf("unexpected type for v1.Secret stringData, got %T", stringDataAny)
```
Comment (tomasaschan) [inline mockkubeapiserver/postresource.go:94]:
```suggestion
				return fmt.Errorf("unexpected type for v1.Secret stringData: expected map[string]any, got %T", stringDataAny)
```

File: mockkubeapiserver/postresource.go [hunk 2]
```diff
@@ -73,8 +74,42 @@ func (req *postResource) Run(ctx context.Context, s *MockKubeAPIServer) error {
...
+			}
+			dataAny, dataFound := obj.Object["data"]
+			if !dataFound {
+				dataAny = make(map[string]any)
+				obj.Object["data"] = dataAny
+			}
+			dataMap, ok := dataAny.(map[string]any)
+			if !ok {
+				return fmt.Errorf("unexpected type for v1.Secret data, got %T", dataAny)
```
Comment (tomasaschan) [inline mockkubeapiserver/postresource.go:103]:
```suggestion
				return fmt.Errorf("unexpected type for v1.Secret data: expected map[string]any, got %T", dataAny)
```

File: mockkubeapiserver/postresource.go [hunk 3]
```diff
@@ -73,8 +74,42 @@ func (req *postResource) Run(ctx context.Context, s *MockKubeAPIServer) error {
...
+			}
+			for k, vAny := range stringDataMap {
+				v, ok := vAny.(string)
+				if !ok {
+					return fmt.Errorf("unexpected type for v1.Secret stringData entry, got %T", vAny)
```
Comment (tomasaschan) [inline mockkubeapiserver/postresource.go:108]:
```suggestion
					return fmt.Errorf("unexpected type for v1.Secret stringData entry %s: expected string, got %T", k, vAny)
```

File: mockkubeapiserver/postresource.go [hunk 4]
```diff
@@ -73,8 +74,42 @@ func (req *postResource) Run(ctx context.Context, s *MockKubeAPIServer) error {
...
+				}
+				dataMap[k] = base64.StdEncoding.EncodeToString([]byte(v))
```
Comment (tomasaschan) [inline mockkubeapiserver/postresource.go:110]: I'm not familiar enough with what Go will do through all of these casts; will this actually end up mutating `obj`? Or does it just mutate the `dataMap` we've created, but not actually propagate those changes back?

**Review by tomasaschan — COMMENTED** [review]
A bit of a nit, but still: the errors could be more descriptive.

Also, wouldn't hurt to cover/document this behavior in tests :)

**Comment by justinsb** [comment]
> Also, wouldn't hurt to cover/document this behavior in tests :)

A great point, I was thinking about that the other day.  This change was to support testing in k8s-config-connector, so it got test coverage there, but we could easily create some simple tests directly in the repo - let me do that!

File: mockkubeapiserver/postresource.go (hunk 2 above)
Comment (justinsb) [reply mockkubeapiserver/postresource.go:103]: Thanks, agree 100%

File: mockkubeapiserver/postresource.go (hunk 4 above)
Comment (justinsb) [reply mockkubeapiserver/postresource.go:110]: So the answer is yes, and the test will (hopefully!) prove it.  I'll have a go at rewriting it to be more self-evidently true though, I agree there's too much happening here.

**Comment by justinsb** [comment]
This is escalating a bit ... looks like I need to break out a new module to avoid a circular dependency: https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/pull/398

**Comment by justinsb** [comment]
Had a bit of trouble with the go.sum / go.mod checking (dev/format-gomod), but I finally figured out that the go.sum difference was to do with my cache; I did `go clean -cache -modcache` before running `dev/format-gomod`.  I think it's still worth trying go.work (and I added a note to the dev/format-gomod script and to the GitHub action error message)

**Comment by justinsb** [comment]
This _might_ be much easier if we merged a (basically empty) `ktest` module first - then I think we could stay away from `go work sync` which is the root of all the problems.  Maybe that's a reasonable compromise that's worth a go.

**Comment by justinsb** [comment]
I tried staging the ktest module in #400.  I think that would mean I could back out the `go work sync` logic which gave me so much trouble

**Comment by justinsb** [comment]
Found a repo that checks in go.work: kubernetes/kubernetes !  I think we're on solid ground :-)

File: ktest/testharness/manifest.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+func ParseObjects(ctx context.Context, manifest string) ([]*unstructured.Unstructured, error) {
```
Comment (yuwenma) [inline ktest/testharness/manifest.go:30]: 2 years passed and we still have to write this. 🤣

File: mockkubeapiserver/postresource.go [hunk 6]
```diff
@@ -73,8 +74,55 @@ func (req *postResource) Run(ctx context.Context, s *MockKubeAPIServer) error {
...
+func beforeObjectCreation(ctx context.Context, obj *unstructured.Unstructured) error {
+	gvk := obj.GroupVersionKind()
+	if gvk == secretGVK {
+		return beforeSecretCreation(ctx, obj)
+	}
+	return nil
+}
+
+func beforeSecretCreation(ctx context.Context, obj *unstructured.Unstructured) error {
+	// If there is any stringData, merge it into data
```
Comment (yuwenma) [inline mockkubeapiserver/postresource.go:102]: ➕

**Review by yuwenma — COMMENTED** [review]
/lgtm
/approve

This PR has so many wonderful pieces and make the code base cleaner and more readable.

**Pull request merged**